litert/vendors/qualcomm/core/transformation/mha_to_sha.cc

```diff
@@ -0,0 +1,373 @@
+#include <functional>
+#include <vector>
+
+#include "litert/vendors/qualcomm/core/builders/concatenation_op_builder.h"
+#include "litert/vendors/qualcomm/core/builders/reshape_op_builder.h"
+#include "litert/vendors/qualcomm/core/builders/split_op_builder.h"
+#include "litert/vendors/qualcomm/core/op_code.h"
+#include "litert/vendors/qualcomm/core/tensor_pool.h"
+#include "litert/vendors/qualcomm/core/utils/log.h"
+#include "litert/vendors/qualcomm/core/wrappers/op_wrapper.h"
+#include "QnnInterface.h"  // from @qairt
+#include "QnnOpDef.h"  // from @qairt
+
+namespace qnn {
+namespace {
+
+// Returns a boolean indicating whether the output tensor of op1 at out_id is
+// connected to the input tensor of op2 at in_id.
+#define IS_CONNECTED(op1, out_id, op2, in_id)     \
+  (ops[start_id + op1].GetOutputTensor(out_id) == \
+   ops[start_id + op2].GetInputTensor(in_id))
+
+constexpr size_t kMulIndex = 0;
+constexpr size_t kTransposePrefillIndex = 1;
+constexpr size_t kReshapePrefillIndex = 2;
+constexpr size_t kMatMulK1Index = 1;
+constexpr size_t kMatMulK2Index = 2;
+constexpr size_t kConcatIndex = 3;
+constexpr size_t kReshape0Index = 4;
+constexpr size_t kAddIndex = 5;
+constexpr size_t kReshape1Index = 6;
+constexpr size_t kSoftmaxIndex = 7;
+constexpr size_t kSlice1Index = 8;
+constexpr size_t kSlice2Index = 9;
+constexpr size_t kMatMulV1Index = 10;
+constexpr size_t kMatMulV2Index = 11;
+constexpr size_t kAdd2Index = 12;
+constexpr size_t kReshape2Index = 13;
+constexpr size_t kTranspose2Index = 14;
+constexpr size_t kReshape3Index = 15;
+
+// Emplaces the operator with updated inputs/outputs into new_ops.
+// This function copies the source_op and updates the tensors according to
+// inputs and outputs. The std::nullopt input/output element indicates that the
+// tensor will not be updated.
+void EmplaceOpWithIO(
+    std::vector<OpWrapper>& new_ops, const OpWrapper& source_op,
+    const std::vector<std::optional<qnn::TensorWrapperRef>>& inputs,
+    const std::vector<std::optional<qnn::TensorWrapperRef>>& outputs) {
+  OpWrapper ret = source_op;
+  ret.UpdateTensors(inputs, outputs);
+  new_ops.emplace_back(ret);
+}
+
+TensorWrapper& BuildSingleSHA(std::vector<OpWrapper>& ops, size_t start_id,
+                              std::vector<OpWrapper>& new_ops,
+                              TensorPool& tensor_pool,
+                              qnn::TensorWrapper& sha_input,
+                              const ::qnn::OpWrapper& first_mul,
+                              size_t num_heads) {
+  // Mul
+  auto& mul_output = tensor_pool.CloneNativeTensorFrom(
+      first_mul.GetOutputTensor(0), sha_input.GetDims());
+  EmplaceOpWithIO(new_ops, first_mul, {sha_input, std::nullopt}, {mul_output});
+  // MatMul 1
+  const auto& matmulk_cache_output =
+      ops[start_id + kMatMulK1Index].GetOutputTensor(0);
+  std::vector<uint32_t> matmul1_output_dim = matmulk_cache_output.GetDims();
+  matmul1_output_dim[2] = matmul1_output_dim[2] / num_heads;
```

Comment:
NIT:
```suggestion
  matmul1_output_dim[2] /= num_heads;
```
Also, they should be all same value, maybe we can store the first one for others.